File: backend/models/habit.go
package models

import (
	"time"
)

// Habit represents a habit in the system
type Habit struct {
	ID                string     `json:"id" db:"id"`
	UserID            string     `json:"user_id" db:"user_id"`
	Name              string     `json:"name" db:"name"`
	Icon              *string    `json:"icon,omitempty" db:"icon"`
	Color             *string    `json:"color,omitempty" db:"color"`
	Frequency         Frequency  `json:"frequency" db:"frequency"`
	Note              *string    `json:"note,omitempty" db:"note"`
	TwoMinuteVersion  *string    `json:"two_minute_version,omitempty" db:"two_minute_version"`
	StackAfterHabitID *string    `json:"stack_after_habit_id,omitempty" db:"stack_after_habit_id"`
	IsNegative        bool       `json:"is_negative" db:"is_negative"`
	Archived          bool       `json:"archived" db:"archived"`
	CreatedAt         time.Time  `json:"created_at" db:"created_at"`
	UpdatedAt         time.Time  `json:"updated_at" db:"updated_at"`

	// Optional joined fields
	StackAfterHabit *Habit         `json:"stack_after_habit,omitempty" db:"-"`
	Checkins        []Checkin      `json:"checkins,omitempty" db:"-"`
	CurrentStreak   int            `json:"current_streak,omitempty" db:"-"`
	Subtasks        []HabitSubtask `json:"subtasks,omitempty" db:"-"`
}

// HabitCreateRequest represents the request for creating a habit
type HabitCreateRequest struct {
	Name              string          `json:"name" binding:"required" validate:"required,max=255"`
	Icon              *string         `json:"icon,omitempty" validate:"omitempty,max=100"`
	Color             *string         `json:"color,omitempty" validate:"omitempty,max=50"`
	Frequency         Frequency       `json:"frequency" binding:"required" validate:"required"`
	Note              *string         `json:"note,omitempty"`
	TwoMinuteVersion  *string         `json:"two_minute_version,omitempty"`
	StackAfterHabitID *string         `json:"stack_after_habit_id,omitempty"`
	IsNegative        bool            `json:"is_negative"`
	Subtasks          []SubtaskRequest `json:"subtasks,omitempty"`
}

// HabitUpdateRequest represents the request for updating a habit
type HabitUpdateRequest struct {
	Name              *string          `json:"name,omitempty"`
	Icon              *string          `json:"icon,omitempty"`
	Color             *string          `json:"color,omitempty"`
	Frequency         *Frequency       `json:"frequency,omitempty"`
	Note              *string          `json:"note,omitempty"`
	TwoMinuteVersion  *string          `json:"two_minute_version,omitempty"`
	StackAfterHabitID *string          `json:"stack_after_habit_id,omitempty"`
	IsNegative        *bool            `json:"is_negative,omitempty"`
	Archived          *bool            `json:"archived,omitempty"`
	Subtasks          []SubtaskRequest `json:"subtasks,omitempty"`
}

// Frequency represents the habit frequency configuration
type Frequency struct {
	Type         string `json:"type"` // "daily", "weekly", "times_per_week", "interval"
	TimesPerDay  int    `json:"times_per_day,omitempty"`
	Days         []int  `json:"days,omitempty"`      // 0-6 for weekly specific days
	Times        int    `json:"times,omitempty"`     // for times_per_week
	EveryNDays   int    `json:"every_n_days,omitempty"` // for interval
}

// TableName returns the database table name for the Habit model
func (Habit) TableName() string {
	return "habits"
}

File: mobile/lib/features/habits/models/habit.dart
class HabitFrequency {
  final String type; // daily, weekly, times_per_week, interval
  final int? timesPerDay;
  final List<int>? days; // 0-6 for weekly specific days
  final int? times; // for times_per_week
  final int? everyNDays; // for interval

  const HabitFrequency({
    required this.type,
    this.timesPerDay,
    this.days,
    this.times,
    this.everyNDays,
  });

  factory HabitFrequency.fromJson(Map<String, dynamic> json) {
    return HabitFrequency(
      type: json['type'] as String? ?? 'daily',
      timesPerDay: json['times_per_day'] as int?,
      days: (json['days'] as List<dynamic>?)?.map((e) => e as int).toList(),
      times: json['times'] as int?,
      everyNDays: json['every_n_days'] as int?,
    );
  }

  Map<String, dynamic> toJson() => {
        'type': type,
        if (timesPerDay != null) 'times_per_day': timesPerDay,
        if (days != null) 'days': days,
        if (times != null) 'times': times,
        if (everyNDays != null) 'every_n_days': everyNDays,
      };

  String get label {
    switch (type) {
      case 'times_per_week':
        return '${times ?? 1}× per week';
      case 'interval':
        return 'Every ${everyNDays ?? 1} day${(everyNDays ?? 1) == 1 ? '' : 's'}';
      case 'weekly':
        if (days != null && days!.isNotEmpty) {
          const dayNames = ['Sun', 'Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat'];
          return days!.map((d) => dayNames[d]).join(', ');
        }
        return 'Weekly';
      default:
        return 'Daily';
    }
  }
}

class HabitSubtask {
  final String id;
  final String habitId;
  final String label;
  final int sortOrder;

  const HabitSubtask({
    required this.id,
    required this.habitId,
    required this.label,
    required this.sortOrder,
  });

  factory HabitSubtask.fromJson(Map<String, dynamic> json) {
    return HabitSubtask(
      id: json['id'] as String,
      habitId: json['habit_id'] as String,
      label: json['label'] as String,
      sortOrder: json['sort_order'] as int? ?? 0,
    );
  }
}

class Habit {
  final String id;
  final String userId;
  final String name;
  final String? icon;
  final String? color;
  final String? note;
  final String? twoMinuteVersion;
  final String? stackAfterHabitId;
  final bool archived;
  final bool isNegative;
  final HabitFrequency frequency;
  final List<HabitSubtask> subtasks;
  final DateTime? createdAt;

  const Habit({
    required this.id,
    required this.userId,
    required this.name,
    this.icon,
    this.color,
    this.note,
    this.twoMinuteVersion,
    this.stackAfterHabitId,
    required this.archived,
    this.isNegative = false,
    required this.frequency,
    this.subtasks = const [],
    this.createdAt,
  });

  factory Habit.fromJson(Map<String, dynamic> json) {
    return Habit(
      id: json['id'] as String,
      userId: json['user_id'] as String,
      name: json['name'] as String,
      icon: json['icon'] as String?,
      color: json['color'] as String?,
      note: json['note'] as String?,
      twoMinuteVersion: json['two_minute_version'] as String?,
      stackAfterHabitId: json['stack_after_habit_id'] as String?,
      archived: json['archived'] as bool? ?? false,
      isNegative: json['is_negative'] as bool? ?? false,
      frequency: json['frequency'] != null
          ? HabitFrequency.fromJson(json['frequency'] as Map<String, dynamic>)
          : const HabitFrequency(type: 'daily'),
      subtasks: (json['subtasks'] as List<dynamic>?)
              ?.map((e) => HabitSubtask.fromJson(e as Map<String, dynamic>))
              .toList() ??
          [],
    );
  }
}
